docs/CONVENTIONS.md

```diff
@@ -19,11 +19,149 @@ There isn't currently a way to make IntelliJ's real-time formatting adhere stric
+
+Our pattern for this is to declare “new” and “existing” model classes, e.g., NewOrganizationModel and ExistingOrganizationModel, where the fields in question are either nullable or don't exist at all on the "new" classes.
+
+We do this either of two ways: by declaring completely separate classes or by using generics and typealiases.
+
+Separate classes are straightforward:
+
+```kotlin
+data class NewOrganizationModel(
+    val name: String
+)
+
+data class ExistingOrganizationModel(
+    val id: OrganizationId,
+    val name: String
+)
+```
+
+The disadvantage of this approach is that it’s awkward to write code that works with both existing and new objects. So another approach is to use one class:
+
+```kotlin
+data class OrganizationModel<ID : OrganizationId?>(
+    val id: ID,
+    val name: String
+)
+
+typealias NewOrganizationModel = OrganizationModel<Nothing?>
+typealias ExistingOrganizationModel = OrganizationModel<OrganizationId>
+```
+
+In this approach, there is always an id field, but it must always be set to null if you’re using the NewOrganizationModel typealias, and can never be null if you’re using ExistingOrganizationModel. Functions that don’t care which variety of model they’re using can accept OrganizationModel<*> parameters.
+
+### DAOs and stores
+
+jOOQ generates a “DAO” class for each table (other than enum tables). The DAO class gives you a basic set of CRUD operations. It works with the Row classes.
+
+We often use the DAO classes for inserting new rows and fetching single rows; they’re more succinct than constructing SQL statements.
+
+However, we usually don’t call them directly: generally, we define a “store” class and call the DAO from the store.
+
+Why we use store classes:
+
+Permission checks. We want to throw an exception if the current user doesn’t have permission to read or modify a particular object in the database.
+
+Populating mandatory fields. Many tables have fields like “created time” or “modified by user” and the store class can ensure that these are always set to the correct values rather than forcing every caller to remember to set them.
+
+Models. The store classes can accept and return models rather than rows.
+
+Sorting. The DAO classes have methods to fetch multiple rows, but they don’t guarantee a sort order.
+
+More sophisticated operations. Sometimes we want to do things that aren’t trivial single-table CRUD operations.
+
+Stores should not depend on other stores! See the next section for more. But it’s fine for a store to depend on DAOs and low-level helper objects.
+
+### Rows vs. Records
+
+TL;DR: Generally prefer the `Row` classes over the `Record` classes. The rest of this section talks about the differences between the two.
```

Comment:
To be clear, this statement "generally prefer the `Row` classes" is how it currently is, but we will likely change in favor of using `Record` in the future. So if I am doing something now, I should use `Record` more often, is that right? 

When I originally read this, I went back to rework something in a current PR, because I am using the `Record`. Since I was just accessing some data as opposed to potentially interacting with the table again, it seemed like I should use the `Row`. But if we are likely to change in favor of using `Record`, should I keep it as is?

Reply:
Correct. We've been using the `Row` classes but they are kind of second-class citizens in jOOQ and the `Record` ones are more full-featured. For now, you should use whichever is more convenient, but at some point I want to come up with a plan to migrate over to `Record`s. The reason I'm not pushing for it too much in new code right now is just that I think it'll make sense to come up with a usage/style guide so we use them consistently everywhere.

Once we've moved over, we can turn off code generation for the `Row` classes and we can get rid of a bunch of `Dao` dependencies all over the place since all you need is a `DSLContext` to read and write `Record`s.